### PR TITLE
feat: substra example

### PR DESCRIPTION
**Pull request by ThibaultFy**
# Summary

Substrafl example to show a way to use dataset from Flamby in local subprocess mode.

The example is a jupyter notebook.
A .py file is provided to simplify the life of the beloved reviewers (and can be removed in the end, or merged with the notebook, no strong opinion).

### CODE REVIEW

File: integration/Substra/assets/dataset/opener_pooled_test.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from flamby.datasets.fed_tcga_brca import FedTcgaBrca
+import substratools as tools
+from torch.utils.data import DataLoader
+
+
+class MnistOpener(tools.Opener):
```
Comment (jeandut) [inline integration/Substra/assets/dataset/opener_pooled_test.py:6]: Remove Mnist from the name

File: integration/Substra/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+
+# SubstraFL example on FLamby TCGA BRCA dataset
+
+This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning
```
Comment (jeandut) [inline integration/Substra/README.md:4]: illustrates/proposes

File: integration/Substra/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+using de Federated Average strategy on the TCGA BRCA dataset, using FLamby.
```
Comment (jeandut) [inline integration/Substra/README.md:5]: the Federated Averaging strategy on the TCGA BRCA dataset with FLamby

File: integration/Substra/assets/dataset/description.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# FLamby TCGA BRCA
+
+This dataset is [the FLamby TCGA BRCA dataset](https://github.com/owkin/FLamby/tree/main/flamby/datasets/fed_tcga_brca).
+
+## Opener usage
+
+The opener exposes 6 methods:
```
Comment (jeandut) [inline integration/Substra/assets/dataset/description.md:7]: Correct the deccription

File: integration/Substra/plot_substrafl_flamby.py [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+# We work with six different organizations, defined by their IDs. Both organizations provide a dataset. One of them
+# will also provide the algorithm and will register the machine learning tasks.
+#
+# Once these variables defined, we can create our Substra :ref:`documentation/references/sdk:Client`.
+#
+# This example runs in local mode, simulating a **federated learning** experiment.
+
+from substra import Client
+
+N_CLIENTS = 6
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:57]: Can be imported from FLamby

File: integration/Substra/plot_substrafl_flamby.py [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+# **datasample**.
+# Concretely, a metric corresponds to an archive *(tar or zip file)*, automatically build
+# from:
+#
+# - a **Python scripts** that implement the metric computation
+# - a `Dockerfile <https://docs.docker.com/engine/reference/builder/>`__ to specify the required dependencies of the
+#   **Python scripts**
+
+
+inputs_metrics = [
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:121]: Could you explain what's going on here ?

File: integration/Substra/plot_substrafl_flamby.py [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+
+# %%
+# Registering dataset
+# ===================
+#
+# A :ref:`documentation/concepts:Dataset` is composed of an **opener**, which is a Python script with the instruction
+# of *how to load the data* from the files in memory, and a **description markdown** file.
+
+
+train_datasets_dict = dict()
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:159]: Why do we need a dict in addition to the list below ?

File: integration/Substra/plot_substrafl_flamby.py [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+
+    # Add the dataset to the client to provide access to the opener in each organization.
+    train_dataset_key = client.add_dataset(train_datasets_dict[org_id])
+    assert train_dataset_key, "Missing data manager key"
+
+    # Add the training data on each organization.
+    data_sample = DataSampleSpec(
+        data_manager_keys=[train_dataset_key],
+        test_only=False,
+        path=pathlib.Path.cwd() / "empty",
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:204]: Could you explain what is this path doing ?

File: integration/Substra/plot_substrafl_flamby.py [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+    train_data_node = TrainDataNode(
+        organization_id=org_id,
+        data_manager_key=train_dataset_key,
+        data_sample_keys=[train_datasample_key],
+    )
+    train_data_nodes.append(train_data_node)
+
+# Add the testing data.
+
+client = clients[ALGO_ORG_ID]
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:221]: There should be 6 centers for testing not 1

File: integration/Substra/plot_substrafl_flamby.py [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+# Number of model update between each FL strategy aggregation.
+NUM_UPDATES = 16
+
+index_generator = NpIndexGenerator(
+    batch_size=batch_size,
+    num_updates=NUM_UPDATES,
+)
+
+
+class TorchDataset:
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:314]: Use inheritance instead of this please

File: integration/Substra/plot_substrafl_flamby.py [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+
+# %%
+# Algo dependencies
+# =================
+#
+# The **dependencies** needed for the :ref:`substrafl_doc/api/algorithms:Torch Algorithms` are specified by a
+# :ref:`substrafl_doc/api/dependency:Dependency` object, in order to install the right library in the Python
+# environment of each organization.
+
+algo_deps = Dependency(pypi_dependencies=["torch==1.12.0"])
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:359]: Why is it the only dep we need ?

**Comment by ThibaultFy** [comment]
@jeandut thanks for all the comments but the PR is not ready for review yet, only in draft for now :)

File: integration/Substra/plot_substrafl_flamby.py (hunk 9 above)
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:221]: For this one, for now in the example we are testing on the pooled test dataset. Do you mean that you would prefer testing the model on each test subset ?

Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:221]: @jeandut up

File: integration/Substra/plot_substrafl_flamby.py (hunk 8 above)
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:204]: It is explain in the comment section

File: integration/Substra/plot_substrafl_flamby.py (hunk 10 above)
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py]: I changed the interaction to make this part more readable. Now, the opener does not return the FLamby dataset but a dict containing the argument to instantiate it. I like it better this way thanks for insisting on this one 😄 

The opener for testing still instantiate a dataset to access to the y values though...

File: integration/Substra/plot_substrafl_flamby.py (hunk 11 above)
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:365]: In fact the dependencies are not used in local mode. I added a comments about this. 
Indeed I'm not sure that this is the only dep we need, I'll check on my side !

File: integration/Substra/plot_substrafl_flamby.py (hunk 6 above)
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:125]: This would ask a lot of effort and and might add confusions to explain everything that is happening here... I will use a Joker for now and discuss it with the squad.

File: integration/Substra/.gitignore [hunk 12]
```diff
@@ -0,0 +1,5 @@
+# Other
+local-worker/
+experiment_summaries/*
+!experiment_summaries/.gitkeep
```
Comment (Esadruhn) [inline integration/Substra/.gitignore:4]: just to share another way of doing this: 

do not put anything in this `.gitignore`
in the `experiment_summaries` folder, add a `.gitignore` file with the following contents:

```
*
!.gitignore
```

File: integration/Substra/.gitignore [hunk 13]
```diff
@@ -0,0 +1,5 @@
...
+*.zip
```
Comment (Esadruhn) [inline integration/Substra/.gitignore:5]: what is the `empty` folder used for?

File: integration/Substra/README.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
+# Substra
+
+[Substra](https://docs.substra.org/en/latest/) is an open source federated learning framework. It provides a flexible python interface and a web app to run federated learning training at scale.
+
+Substra main usage is a production one. It has already been deployed and used by hospitals and biotech companies (see the [MELLODDY](https://www.melloddy.eu/) project for instance). Yet Substra can also be used on a single machine on a virtually splitted dataset for two use cases:
+
+debugging code before launching experiments on a real network
+
+performing FL simulations
```
Comment (Esadruhn) [inline integration/Substra/README.md:9]:
```suggestion
- debugging code before launching experiments on a real network
- performing FL simulations
```

File: integration/Substra/README.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
+# Substra
+
+[Substra](https://docs.substra.org/en/latest/) is an open source federated learning framework. It provides a flexible python interface and a web app to run federated learning training at scale.
```
Comment (Esadruhn) [inline integration/Substra/README.md:3]:
```suggestion
[Substra](https://docs.substra.org/) is an open source federated learning framework. It provides a flexible python interface and a web app to run federated learning training at scale.
```

`latest` does not match any release, let's not redirect users to it
I expect this link to be functional in the next week

File: integration/Substra/README.md [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+# SubstraFL example on FLamby TCGA BRCA dataset
+
+This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning
+using de Federated Average strategy on the TCGA BRCA dataset.
+
+The objective of this example is to launch a *federated learning* experiment on the six centers provided on Flamby
+(called organization on Substra), using the **FedAvg strategy** on the **baseline model**.
+
+This example does not use the deployed platform of Substra and will run in local mode.
```
Comment (Esadruhn) [inline integration/Substra/README.md:21]:
```suggestion
This example runs in local mode, no need to deploy a Substra platform to run it.
```

File: integration/Substra/README.md [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+## Requirements
+
+- To run this example locally, please make sure to download and unzip in the same directory as this example the
+    assets needed to run it:
+
+    Please ensure to have all the libraries installed, a *requirements.txt* file is included in the zip file, where
+    you can run the command: `pip install -r requirements.txt` to install them.
```
Comment (Esadruhn) [inline integration/Substra/README.md:29]: no zip folder here, you can directly say to do `pip install -r requirements.txt`

File: integration/Substra/README.md [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+## Objective
+
+This example will run a federated training on all 6 centers.
```
Comment (Esadruhn) [inline integration/Substra/README.md:33]:
```suggestion
This example runs a federated training on all 6 centers.
```

**Review by Esadruhn — CHANGES_REQUESTED** [review]
Thanks for the PR!
I did not run the example, minor comments

File: integration/Substra/.gitignore (hunk 13 above)
Comment (ThibaultFy) [reply integration/Substra/.gitignore:5]: It is to point a folder for the datasamples, as it is a required arg but we have no data to load in a FLamby setup

File: integration/Substra/README.md (hunk 15 above)
Comment (ThibaultFy) [reply integration/Substra/README.md:3]: Yes, I pointed it out in the PR description. I'm waiting for the release to fix the doc link

File: integration/Substra/README.md (hunk 17 above)
Comment (ThibaultFy) [reply integration/Substra/README.md:28]: Copy past too fast, thanks for the catch !

File: integration/Substra/assets/dataset/opener_pooled_test.py [hunk 19]
```diff
@@ -0,0 +1,20 @@
+from flamby.datasets.fed_tcga_brca import FedTcgaBrca
+import substratools as tools
+from torch.utils.data import DataLoader
+
+
+class FlambyTestOpener(tools.Opener):
+    def get_X(self, folders):
+        config = {"pooled": True, "train": False}
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/assets/dataset/opener_pooled_test.py:8]: To load the test samples instead of the train ones

File: integration/Substra/plot_substrafl_flamby.py [hunk 20]
```diff
@@ -0,0 +1,438 @@
...
+#
+# Now that all our [**Assets**](https://docs.substra.org/en/latest/documentation/concepts.html#assets) are well defined, we can create the [**TrainDataNodes**](https://docs.substra.org/en/latest/substrafl_doc/api/nodes.html#traindatanode) and the [**TestDataNodes**](https://docs.substra.org/en/latest/substrafl_doc/api/nodes.html#testdatanode) to gathered the [**Datasets**](https://docs.substra.org/en/latest/documentation/concepts.html#dataset) and the **datasamples** on the specified organization.
+
+# In[8]:
+
+
+train_data_nodes = list()
+test_data_nodes = list()
+
+empty_path = pathlib.Path.cwd() / "empty"
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:193]: good idead

File: integration/Substra/plot_substrafl_flamby.py [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+        predictions = torch.Tensor([])
+        with torch.inference_mode():
+            for x, _ in predict_loader:
+                x = x.to(self._device)
+                predictions = torch.cat((predictions, self._model(x)), 0)
+
+        predictions = predictions.cpu().detach()
+        self._save_predictions(predictions, predictions_path)
+
+        return predictions
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:350]: In the default local predict, we assume that the __getitem__ of the dataset take into account the 'is_inference' argument and only return x. But as the __getitem__ is the one from flamby, the is_inference in used here. So we need to overwrite the _local_predict just for `for x, _ in predict_loader:` instead of `for x in predict_loader:`

File: integration/Substra/plot_substrafl_flamby.py [hunk 22]
```diff
@@ -0,0 +1,438 @@
...
+
+# In[8]:
+
+
+train_data_nodes = list()
+test_data_nodes = list()
+
+empty_path = pathlib.Path.cwd() / "empty"
+
+for ind, org_id in enumerate(ORGS_ID):
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:195]: It simplifies the code I think, thanks

File: integration/Substra/plot_substrafl_flamby.py [hunk 23]
```diff
@@ -0,0 +1,438 @@
...
+
+# In[9]:
+
+
+from flamby.datasets.fed_tcga_brca import (
+    Baseline as baseline_model,
+    BaselineLoss as baseline_loss,
+    Optimizer as optimizer,
+    BATCH_SIZE as batch_size,
+    LR as ixi_lr,
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:268]: It was a copy past from another FLamby example. I agree that's its a bit better without the renaming. @jeandut I let you decide 😄

File: integration/Substra/plot_substrafl_flamby.py [hunk 24]
```diff
@@ -0,0 +1,438 @@
...
+# In[11]:
+
+
+NUM_UPDATES = 16
+
+index_generator = NpIndexGenerator(
+    batch_size=batch_size,
+    num_updates=NUM_UPDATES,
+)
+from flamby.datasets.fed_tcga_brca import FedTcgaBrca
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py:316]: Yes of course, legacy from some tests, thanks for the catch

File: integration/Substra/assets/metric/Dockerfile [hunk 25]
```diff
@@ -0,0 +1,11 @@
+# this base image works in both CPU and GPU enabled environments
+FROM ghcr.io/substra/substra-tools:0.15.0-nvidiacuda11.6.0-base-ubuntu20.04-python3.9
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/assets/metric/Dockerfile]: Thanks! I corrected this, and I just tried to run in docker but in fact it's not possible (or at least not straightforward) because flamby is a dependency of the metric, but can't be installed with pip. It has to be installed in editable mode to access the datasets...

File: integration/Substra/plot_substrafl_flamby.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (Esadruhn) [reply integration/Substra/plot_substrafl_flamby.py:343]: Thank you for the explanation!
Can you add a small comment to explain this?

File: integration/Substra/assets/metric/Dockerfile (hunk 25 above)
[the remark this answers is not in the document]
Comment (Esadruhn) [reply integration/Substra/assets/metric/Dockerfile]: then you can either temporarily copy the source code to the metric folder, then use the `COPY` command to add it to the Docker image, then run `RUN pip install -e .`

or try to git clone the repo from inside the Docker image and install it

And I think you also need to add it to the substrafl algo local dependencies

in all cases it seems like it would take a long time to build and run in Docker mode, maybe we can explain in the README that this example runs in local subprocess mode but not in Docker mode and say a few words why

File: integration/Substra/plot_substrafl_flamby.py [hunk 26]
```diff
@@ -0,0 +1,438 @@
...
+# This example runs in local mode, simulating a **federated learning** experiment.
+
+# # Client and data preparation
+# ## Imports
+
+# In[1]:
+
+
+# Choose the subprocess mode to locally simulate the FL process
+get_ipython().run_line_magic("set_env", "DEBUG_SPAWNER = subprocess")
```
[the remark this answers is not in the document]
Comment (Esadruhn) [reply integration/Substra/plot_substrafl_flamby.py]: It does work in notebooks, I tested it and it was used in the product demo
what was the code that did not work?

[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py]: It works indeed... sorry for that, I don't know what I did to conclude that 🙄

File: integration/Substra/assets/metric/Dockerfile (hunk 25 above)
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply integration/Substra/assets/metric/Dockerfile]: I updated the readme for now. I think there is a problem in setting FLamby as a local dependency  because of the local worker folder.

File: integration/Substra/plot_substrafl_flamby.py (hunk 9 above)
Comment (jeandut) [reply integration/Substra/plot_substrafl_flamby.py:211]: Yes the proper way to test would be on each test set for each center.

Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py]: The example have been updated, ready for a second review @jeandut :)

File: integration/Substra/README.md [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+Substra main usage is a production one. It has already been deployed and used by hospitals and biotech companies (see the [MELLODDY](https://www.melloddy.eu/) project for instance). Yet Substra can also be used on a single machine on a virtually splitted dataset for two use cases:
+
+- debugging code before launching experiments on a real network
+- performing FL simulations
+
+Substra was created by [Owkin](https://owkin.com/) and is now hosted by the [Linux Foundation for AI and Data](https://lfaidata.foundation/).
+
+# SubstraFL example on FLamby TCGA BRCA dataset
+
+This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning using de Federated Average strategy on the TCGA BRCA dataset.
```
Comment (jeandut) [inline integration/Substra/README.md:14]:
```suggestion
This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning using the Federated Averaging strategy on the TCGA BRCA dataset.
```

File: integration/Substra/README.md [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+
+The objective of this example is to launch a *federated learning* experiment on the six centers provided on Flamby (called organization on Substra), using the **FedAvg strategy** on the **baseline model**.
```
Comment (jeandut) [inline integration/Substra/README.md:16]:
```suggestion
The objective of this example is to launch a *federated learning* experiment on the six centers provided on Flamby (called organizations in Substra), using the **FedAvg strategy** on the **baseline model**.
```

File: integration/Substra/README.md [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+
+This example runs in local mode, no need to deploy a Substra platform to run it.
```
Comment (jeandut) [inline integration/Substra/README.md:18]:
```suggestion
This example runs in local mode in python runtime, but can also be run on a deployed substra network.
```

File: integration/Substra/README.md [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+
+This example is designed to work in local subprocess mode, and not in local Docker mode. As FLamby is a dependency for both the metric and the algorithm, it would need to be installed in editable mode in each container, which has not been added to the example yet.
+
+## Requirements
+
+Please ensure to have all the libraries installed, a *requirements.txt* file is included in the example, where you can run the command: `pip install -r requirements.txt` to install them.
```
Comment (jeandut) [inline integration/Substra/README.md:24]:
```suggestion
Please ensure to have, in addition to flamby, all the libraries needed for this example installed, a *requirements.txt* file is included in the example, this way you can run the command: `pip install -r requirements.txt` to install them.
```

File: integration/Substra/README.md [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+## Objective
+
+This example runs a federated training on all 6 centers.
+
+This example shows how to interface Substra with Flamby. The modifications to use this example on other dataset are minimal and have to be made on the openers files, the metric file and the main script.
+This example runs in local mode, simulating a **federated learning** experiment.
+
+## Run the example
+
+To run the example, launch `jupyter notebook` and run the cells of the `plot_substrafl_flamby.ipynb`
```
Comment (jeandut) [inline integration/Substra/README.md:35]: Can we ditch the notebook in favor of markdown files to keep homogeneity ?

File: integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py [hunk 32]
```diff
@@ -0,0 +1,24 @@
+from flamby.datasets.fed_tcga_brca import FedTcgaBrca
```
Comment (jeandut) [inline integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py:1]: @Esadruhn @ThibaultFy would it be possible to add class inheritance ? Aka can we pass a base class to the openers's dockerfile somehow ? This would allow to factorize all openers real easily.

Comment (jeandut) [reply integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py:1]: I am wondering also if we can do that if it wouldn't be better to find a way to write only one opener and to dynamically send the appropriate instance of it when running the script (add the center number only or train/test) but maybe it's too much and ends up being even more confusing than having them in clear. What do you think @Esadruhn @ThibaultFy ?

File: integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+    def get_y(self, folders):
+        config = self.get_config()
+        dataset = FedTcgaBrca(**config)
+        dataloader = DataLoader(dataset, batch_size=len(dataset))
+        return next(iter(dataloader))[1]
+
+    def get_config(self):
+        return {"center": 0, "train": False}
+
+    def fake_X(self, n_samples=None):
```
Comment (jeandut) [inline integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py:20]: This should be implemented or deprecated in substra. This example should be adaptable to deployed network as well.

File: integration/Substra/assets/metric/metrics.py [hunk 34]
```diff
@@ -0,0 +1,19 @@
+import substratools as tools
```
Comment (jeandut) [inline integration/Substra/assets/metric/metrics.py:1]: Nice !

File: integration/Substra/plot_substrafl_flamby.ipynb [hunk 35]
```diff
@@ -0,0 +1,684 @@
+{
```
[the remark this answers is not in the document]
Comment (jeandut) [reply integration/Substra/plot_substrafl_flamby.ipynb:1]: As commented let's try to be homogeneous and provide a markdown example for it for now. It can then be changed on substra's side.

File: integration/Substra/plot_substrafl_flamby.py (hunk 6 above)
Comment (jeandut) [reply integration/Substra/plot_substrafl_flamby.py:121]: I am still overly confused by this code, but this should be dealt with on the substrafl's side.

File: integration/Substra/plot_substrafl_flamby.py [hunk 36]
```diff
@@ -0,0 +1,446 @@
...
+            seed=SEED,
+        )
+
+    def _local_predict(self, predict_dataset: torch.utils.data.Dataset, predictions_path):
+
+        predict_loader = torch.utils.data.DataLoader(predict_dataset, batch_size=self._index_generator.batch_size)
+
+        self._model.eval()
+
+        predictions = torch.Tensor([])
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:340]: predictions should be instantiated as a `(len(predict_loader.dataset), nouttput_dim)` object already not concatenated on the go as this is inefficient.

File: integration/Substra/plot_substrafl_flamby.py [hunk 37]
```diff
@@ -0,0 +1,446 @@
+#!/usr/bin/env python
```
Comment (jeandut) [inline integration/Substra/plot_substrafl_flamby.py:1]: Can you:
- convert this file to markdown
- change its name to `flamby-integration-into-substra.md` to comply with flamby's requirements ?

File: integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py (hunk 32 above)
Comment (ThibaultFy) [reply integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py:1]: An other way to do it is to have config json files that we load as datasamples instead of giving empty paths, and to load the config into a single opener. 
I found that bypassing the real behaviour of datasamples to pass config files was not the spirit. we are looking for in a documentation example. But I also agree that it'll remove some complexity to the example...

File: integration/Substra/README.md (hunk 31 above)
Comment (ThibaultFy) [reply integration/Substra/README.md]: Do we want to keep both like febiomed or only keep the markdown ?

File: integration/Substra/README.md (hunk 29 above)
Comment (ThibaultFy) [reply integration/Substra/README.md:18]: As flamby is not easily instalable with pip for now, the example does not support the docker mode nor the deployed mode. We might want to update the example later for both these mode?

File: integration/Substra/plot_substrafl_flamby.py (hunk 36 above)
Comment (ThibaultFy) [reply integration/Substra/plot_substrafl_flamby.py]: To be investigated but out of scope of the example as it is the substrafl implementation.

File: integration/Substra/flamby-integration-into-substra.ipynb [hunk 38]
```diff
@@ -0,0 +1,660 @@
+{
```
Comment (jeandut) [inline integration/Substra/flamby-integration-into-substra.ipynb:1]: remove the notebook please (and from all commits with interactive rebasing please)

Comment (ThibaultFy) [reply integration/Substra/flamby-integration-into-substra.ipynb:1]: Why removing it from all commit ? And remove the ipynb means we remove the source file that generate the md document... You're sure about removing it ?

File: integration/Substra/flamby-integration-into-substra.md [hunk 39]
```diff
@@ -0,0 +1,427 @@
...
+
+- debugging code before launching experiments on a real network
+- performing FL simulations
+
+Substra was created by [**Owkin**](https://owkin.com/) and is now hosted by the [Linux Foundation for AI and Data](https://lfaidata.foundation/).
+
+# SubstraFL example on FLamby TCGA BRCA dataset
+
+This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning
+using de Federated Average strategy on the TCGA BRCA dataset.
```
Comment (jeandut) [inline integration/Substra/flamby-integration-into-substra.md:17]:
```suggestion
using the Federated Averaging strategy on the TCGA BRCA dataset.
```

**Review by Esadruhn — APPROVED** [review]
Good for me, great example!
Were you able to test that it runs with Substra 0.21.0?

File: integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py (hunk 32 above)
Comment (Esadruhn) [reply integration/Substra/assets/dataset/opener_test_MyOrg1MSP.py:1]: I think that one opener per center is in the spirit of Substra, with the paradigm that we have specific loaders per centers, and a common algorithm.
As you say, we could dynamically write the opener from the code, I am afraid this would bring much complexity for the end user

File: integration/Substra/assets/metric/Dockerfile (hunk 25 above)
[the remark this answers is not in the document]
Comment (Esadruhn) [reply integration/Substra/assets/metric/Dockerfile]: ok, let's solve the local Docker mode and deployed mode in a separate PR

File: integration/Substra/flamby-integration-into-substra.ipynb (hunk 38 above)
Comment (Esadruhn) [reply integration/Substra/flamby-integration-into-substra.ipynb:1]: if @ThibaultFy removes it then squashes the commits before merging, it will not appear in the git history anymore, is it enough?

**Review by jeandut — COMMENTED** [review]
Sorry @ThibaultFy but can you change the plot again, the name is Fed-TCGA-BRCA instead of Heart-Disease (my bad but it's the one you chose).
MyOrgs should be replaced by Client idx or Client idx (MyOrgWhatever) and test set can be factorized as the title of the legend block.
Sorry for nitpicking.

File: integration/Substra/plot_substrafl_flamby.py (hunk 36 above)
Comment (jeandut) [reply integration/Substra/plot_substrafl_flamby.py]: I don't understand your answer.
The current implementation is terribly inefficient and the modification should take less than two lines.
This has no connexion with the way it is implemented in substrafl, no ?

File: integration/Substra/flamby-integration-into-substra.ipynb (hunk 38 above)
Comment (jeandut) [reply integration/Substra/flamby-integration-into-substra.ipynb]: Removing in order to get rid of the heavy diffs induced by the notebooks, which take a lot of space.

File: integration/Substra/plot_substrafl_flamby.py (hunk 36 above)
Comment (jeandut) [reply integration/Substra/plot_substrafl_flamby.py]: To be more precise the user is invited to overwrite the local predict method of the algo therefore she or he should be able to use efficient implementations as they please.

File: integration/Substra/flamby-integration-into-substra.md [hunk 40]
```diff
@@ -0,0 +1,465 @@
...
+for i, id in enumerate(ORGS_ID):
+    df = performance_df.query(f"worker == '{id}'")
+    plt.plot(df["round_idx"], df["performance"], label=f"Client {i} ({id})")
+
+plt.legend(loc=(1.1, 0.3), title="Test set")
+plt.show()
+```
+
+![png](markdown-image/output_28_0.png)
+
```
Comment (jeandut) [inline integration/Substra/flamby-integration-into-substra.md:417]:
```suggestion
The last test client has only one event, which reduces the number of admissible pairs to be formed in the concordance index computation, which explains to some extents the flat line.
```

**Review by jeandut — APPROVED** [review]
LGTM, great work @ThibaultFy !
